File: src/com/dslplatform/compiler/client/Either.java
package com.dslplatform.compiler.client;

public class Either<T> {
	private final T value;
	private final String error;

	private Either(T value, String error) {
		this.value = value;
		this.error = error;
	}

	public boolean isSuccess() {
		return error == null;
	}

	public T get() {
		return value;
	}

	public String whyNot() {
		return error;
	}

	public static <S> Either<S> success(S value) {
		return new Either<S>(value, null);
	}

	public static <F> Either<F> fail(String error) {
		return new Either<F>(null, error);
	}
}

File: VisualStudioPlugin/Shared/Either.cs
using System;
using System.Threading.Tasks;

namespace DSLPlatform
{
	public struct Either<T>
	{
		public bool Success { get { return Error == null; } }
		public string Error;
		public T Value;

		public static Either<T> Fail(string msg) { return new Either<T> { Error = msg }; }
	}

	public static class Either
	{
		public static Either<T> Success<T>(T value) { return new Either<T> { Value = value }; }

		public static void OnSuccess<T>(this Task<Either<T>> task, Action<T> continueWith)
		{
			task.ContinueWith(t =>
			{
				if (t.Exception == null && t.Result.Success)
					continueWith(t.Result.Value);
			});
		}
	}

}
